### DISCUSSION TITLE
[Feature Request] Relationship CRUDs inside Show Operation (or as a Widget)

### DISCUSSION THREAD

**Discussion by sergioxdev** [Laravel-Backpack/community-forum#262] [General]
Hi Guys

I put my question to you, I have already realized this functionality with version 3.5, and I would like to apply it to version 4, and make it available to everyone.

the goal is to create a Show page that displays in addition to the fields of the X model, the lists of the associated Y and Z models.

example:
category books, I access the show page that shows me the name of the category "comedy", under add the list of all the comedy books. The list will be the same as the view of the index () with all the actions (preview, edit, delete, [by clicking on edit it will be possible to edit the book through a modal window])).

I hope I explained the question well, because I think it's a very useful feature.
for any clarification just ask.

**Comment by pxpm** [comment]
Hello @xdasty 

Hope to find you well. 

Indeed this (or similar) is one of the most requested CRUD features along with the **multiply/matrix** field.

Atm we are providing this functionality with a package developed by one of our users [backpack instant fields](https://github.com/webfactor/laravel-backpack-instant-fields)

It's not exactly what you describe with the related index view, but you could see by example how they are doing it and apply to your scenario ?

If you come up with something you think might be integrated in CRUD, please be our guest and submit a PR. Atleast i can promisse you we will discuss it and later on the road see if it fits. 

Best,
Pedro

**Comment by sergioxdev** [comment]
Hi @pxpm 

I attach an image to show the result

![example_show_page](https://user-images.githubusercontent.com/32037075/68875063-7e637e00-0702-11ea-847c-0c12a3a6a651.jpg)

this shows only a relationship, (modulos-> elements),
can be applied to N relationships. the view can also be in tabs.

In this example
I should have access to the element (list) object, from the Modulos controller show

**Comment by lotarbo** [comment]
@xdasty hello, i like how it looks. Nice job. Mb u can make it like separate package.

**Comment by sergioxdev** [comment]
Hi @lotarbo 
as I wrote before, I already made it for version 3.5; but I would like to port for version 4 and make it public.

**Comment by pxpm** [comment]
Hi the @xdasty 

I also like it, especially the concept :)

If you add an element on that page it automatically gets attached to the Modulos or you choose which Modulo to attach when creating ? Maybe you can work with some session data there to pre-populate Modulos select if not doing that.

How about other types of relations ? Poly etc ?

Best,
Pedro

**Comment by sergioxdev** [comment]
Hi @pxpm 

when I add or modify an element, I can:
or take the id of the module directly from the session
or add a hide field with the modulo id.

**Comment by sergioxdev** [comment]
hi guys i'm going crazy

I update you, for version 3.6 I had implemented the solution
I simplified the code a bit, obviously there will be foreach for the various relationships

on the model controller
```php
public function show($id)
{
....
$this->crud->sub_entry  = new ElementCrudController();
$this->crud->sub_entry_name_route= $this->crud->entity_name;
$entry_ = $this->data['entry'];
$this->crud->sub_entry_relation_id= $entry_->getAttributes()['id'];
+ other variables
....
}
```

on the element controller
```php
public function getCrud($crud_name_route_, $crud_relation_id_)
{
	$this->setup();
	$this->crud->addClause('where', $crud_name_route_.'s_id', $crud_relation_id_);
	return $this->crud;
}
````

in show.blade.php
```php
@if(isset($crud->sub_entry))
$getCrud = $crud->sub_entry->getCrud($crud->sub_entry_name_route, $crud->sub_entry_relation_id);
$entries = $getCrud->getEntries();

@include('admin/show_sub_content', ['crud' => $getCrud , 'entries' => $entries, ... other variables...)
```


applying the same principle to version 4
the getCrud function present on the element controller
at the line $this->setup();

returns me
```
Call to a member function setModel() on null (View: Project\resources\views\vendor\backpack\crud\show.blade.php) 
```

**Comment by tabacitu** [comment]
Hi guys,

I think this is a really neat concept @xdasty - great work. I think a great first step towards making the admin panels more interactive would be to have links to the related entries in the Show operation, that's why in the main repository I'd rather take care of [that issue](https://github.com/Laravel-Backpack/CRUD/issues/1662) first. Since that's a little easier to do, I think.

But I'm sure many people would love it just like you've done it - as additional CRUDs inside the Show operation. This is a great second step!

**Perhaps this can be built as a [Backpack Widget](https://backpackforlaravel.com/docs/4.0/base-widgets)?** That way the entire logic could live in one ```crud``` widget file, that the developer can add to their Show operation if they want it? Want more CRUDs inside your Show operation? Add the ```crud``` widget multiple times, and pass a different entity as the parameter. Plus, creating this as a Widget would have the added benefit of being able to use this in other places than the Show operation - for example in the dashboard, custom views, etc.

However you decide to do this - I don't expect it to be easy to build. I see you've already posted a screenshot of it, so kudos to you! Congrats!

(1) Do the line buttons work? 
(2) Does searching & sorting the CRUD work? 
(3) The ```setModel()``` error you mentioned should be because you've called ```setup()``` - I don't think you need to do that in v4. It's called automatically by the constructor.

Cheers!

**Comment by mustafa-online** [comment]
Any progress on this for v4?

**Comment by sergioxdev** [comment]
Hi guys, I had to fix version 3.6, by the end of the month I continue porting to version 4.

**Comment by dividy** [comment]
How's it going?

**Comment by mustafa-online** [comment]
Waiting for this feature ..

**Comment by sergioxdev** [comment]
Hi Guys
sorry for my silence, since we are at home because of the coronavirus, 
I am fixing the code that will be working for version 3.6

I have entered a fairly complete example of how it works
I am attaching the link to the code
https://github.com/sergioxdev/Laravel-Backpack-Relationship-CRUDs-inside-Show-Operation
You can already check and download the code

I will provide a guide to be able to integrate it.
Have a little patience

**Comment by enefeka** [comment]
Hi! Any news on this? Just started working with backpack a month ago and, just like a lot of us, I reeeeally need this feature. Thanks!

**Comment by sergioxdev** [comment]
@enefeka 

For version 3 you can use the code I wrote,
I also uploaded a complete example and it works,
tested for months

looking at the code it is very easy to add

**Comment by enefeka** [comment]
@sergioxdev 

thanks for the reply! I was talking indeed about the v4, since that's what I'm using

**Comment by tabacitu** [comment]
@enefeka we're not actively working on this right now. We're currently knee-deep in bug fixing in order to [get version 4.1 out the door](https://github.com/Laravel-Backpack/CRUD/projects/8), which was a HUGE endeavour. Only afterwards will we be able to prioritize and maybe put together a solution as a non-breaking update.

If you want this sooner, it shouldn't be terribly difficult to [create in 4.0 using Widgets](https://backpackforlaravel.com/docs/4.1/base-widgets#creating-a-custom-widget-type). Or bug @izica to get his code - I believe [that solution is a Backpack column](https://github.com/Laravel-Backpack/CRUD/issues/2744). If you end up with something general, please consider sharing your code so others can benefit too.

Sorry I can't be of more help right now!

**Comment by tabacitu** [comment]
Actually maybe I can. In Backpack 4.1 there are a few alternatives to this:
- You can [add links to your related items](https://backpackforlaravel.com/docs/4.1/release-notes#new-columns-feature-wrapper), that take the admin to that related entry's Show page.
- (or)
- There's [a new "relationship_count" column](https://backpackforlaravel.com/docs/4.1/release-notes#new-column-relationship_count), that can take the admin to the List view of that related item; combine this with filters, and they'd be able to click "22 items" and then see a filtered table with only the related items;

If that works for you, you can [upgrade to 4.1 right](https://backpackforlaravel.com/docs/4.1/upgrade-guide) now, it's super-stable. The only bugs left to squash are related to the `repeatable` field so you wouldn't be affected. Otherwise it's a very stable build - we're using it ourselves for client projects already.

**Comment by enefeka** [comment]
Hi @tabacitu. I might just do both of those things, already asked @izica for the code and I'm now looking at 4.1 and what you said, it also might work for me.

Thanks and keep up that amazing work!

**Comment by sergioxdev** [comment]
Hi Guys

I hope before August to make my work compatible with the latest version,
and I will publish it here.

**Reply by ingeday** [reply]
Hoping this feature!!

**Reply by Saif-XI-Coderz** [reply]
So looking forward to this xD

**Reply by egyjs** [reply]
any update on this? ^_^

**Reply by ASVVIZIT** [reply]
Hey! How is it going? I would like this functionality on 4.1

**Reply by sergioxdev** [reply]
for now I am a bit busy with different jobs, I hope to be able to finish it by the end of the year (for version 4.1)

**Comment by KevinCassier** [comment]
I think this widget is what you are looking for no? https://github.com/izica/relations-widgets-for-backpack

**Comment by ziming** [comment]
Personally I think it would be great if the relations can be organised into tab groups as some model have lots of relations

[Relation A, B, C]

[Relation D, E, F]

[Relation G, H, I]

**Comment by jcastroa87** [comment]
Hello everyone,

Thanks for sharing this!

The good news is that during the development of Backpack v7, we plan to work on this feature to enable displaying related model lists on Show pages with inline actions. While we can’t provide an exact timeline yet—there’s a lot of work lined up—it is indeed on our backlog. We share your enthusiasm for making this kind of functionality more accessible to everyone.

Thank you again for bringing it up and for your past contributions to Backpack! If you have any tips or insights from your v3.5 implementation, feel free to share; we’d love to learn more.

Cheers.
